### DISCUSSION TITLE
当博客部署在github pages时，如何在仓库名不为username.github.io的仓库中正确部署？

### DISCUSSION THREAD

**Discussion by Shinya754** [YunYouJun/valaxy#377] [Q&A] [answered]
我尝试将valaxy博客部署在我的仓库[“re-blog"](https://github.com/Shinya754/re-blog/tree/gh-pages)中，我不想使用

> shinya754.github.io

的原因是这个仓库已经有一个博客，之前使用hexo部署，但是我硬盘损坏，且没有备份博客文件，导致博客文件永久丢失，因此我不太想再去动shinya754.github.io这个仓库。我尝试将博客部署在**re-blog**中，并且设置github pages部署链接为[https://shinya754.github.io/re-blog/](https://shinya754.github.io/re-blog/)，但是按照valaxy文档中的步骤部署之后，得到以下图示结果：
![错误截图](https://s21.ax1x.com/2024/04/12/pFX7NfU.png)
从浏览器控制台报错中可以看到，浏览器在向**shinya754.github.io**请求资源（当然这不是我手动设置的），但是它显然请求不到。

因此，我想知道，是否有什么方法能够修改它请求资源的路径？

或者，如果这种操作是不允许的，只能将博客部署在username.github.io的话，也请告诉我这个情况，非常感谢！

**Comment by YunYouJun** [comment]
设置配置中的 `siteConfig.url` 为你的对应链接。

---

另外也可以使用 Cloudflare / Vercel / Netlify 等工具直接托管站点。

**Reply by Shinya754** [reply]
好的，非常感谢您的解答！

**Reply by Shinya754** [reply]
@YunYouJun 
作者您好，我尝试了您提供的方案

> 设置配置中的 siteConfig.url 为你的对应链接。

但是valaxy依然从**username.github.io**/获取css等文件。


我也尝试将博客部署在gitee上面并使用gitee pages，而这次其获取资源的路径变为**username.gitee.io**/（如username.gitee.io/assets/xxxxx.css），即使我将**siteConfig.url**设置为**username.gitee.io/blog**/。siteConfig应该是有在正确工作，因为我本地部署时，其中的其他配置都能够正确地体现在页面上。在由valaxy自带github部署脚本所生成的**index.html**文件中，js或者css文件的获取路径是以斜杠开始的相对路径，如下：
```html
<script type="module" async crossorigin src="/assets/app-CqKlUxKa.js"></script>
```
这似乎表示这些资源文件会从网站根目录，即**username.github.io**/取得，而非**username.github.io/repo-name**/

我在想valaxy是否可以添加一个选项以支持这种情况呢？

如果我对于valaxy或者web技术的理解有误，也请您不吝赐教，感谢！

**Comment by fanghsiu** [comment, marked as answer]
吾有三策：上氪金，中改名，下配置。(～￣▽￣)～
1. 花钱买个域名
2. 改仓库名 `Shinya754.github.io -> hexo`, `re-blog -> Shinya754.github.io`
3. 如下，valaxy 基于 vite，但注意 public 下的文件不受此设置影响。https://vitejs.cn/vite3-cn/config/shared-options.html#base
https://vitejs.cn/vite3-cn/config/shared-options.html#publicdir
```
// valaxy.config.ts
vite: {
  base: '/re-blog/'
}
```

**Reply by Shinya754** [reply]
酷。感谢你提供的帮助！这确实有效。

**Comment by S-tanley** [comment]
Hi～，我想问一下，我现在在用hexo作为我的blog网站，我现在想要把https://username.github.io作为我的主页，然后https://username.github.io/blog作为我的blog网站。但是当我把我的username.github.io repo里的东西换到blog repo里面之后一直不好用。请问你知道为什么吗。

**Comment by doudou0720** [comment]
@YunYouJun 这个问题解决了吗？

我正在用readthisdocs构建，这个的baseurl是动态变化的，比如 `/latest/`，vitepress是通过动态获取环境变量进行拼接的...


`.vitepress/config.js`
```
 import { defineConfig } from 'vitepress'

 // https://vitepress.dev/reference/site-config
 export default defineConfig({
     // Use Canonical URL, but only the path and with no trailing /
     // End result is like: `/en/latest`
     base: process.env.READTHEDOCS_CANONICAL_URL
     ? new URL(process.env.READTHEDOCS_CANONICAL_URL).pathname.replace(/\/$/, "")
     : "",

 title: "My Awesome Project",
 description: "A VitePress Site",
 })
```

**Reply by YunYouJun** [reply]
可以。当前不需要新增 Valaxy 选项，`valaxy.config.ts` 会在构建时执行，可以直接读取 Read the Docs 提供的环境变量：

```ts
import process from 'node:process'
import { defineValaxyConfig } from 'valaxy'

const canonicalUrl = process.env.READTHEDOCS_CANONICAL_URL
const base = canonicalUrl
  ? `${new URL(canonicalUrl).pathname.replace(/\/+$/, '')}/`
  : '/'

export default defineValaxyConfig({
  siteConfig: {
    url: canonicalUrl || '',
  },
  vite: {
    base,
  },
})
```

我在当前 `main`（Valaxy v1.0.0-rc.3）上用

```bash
READTHEDOCS_CANONICAL_URL=https://example.readthedocs.io/en/latest/ valaxy build --ssg
```

做了完整构建验证，产物符合预期：

- JS/CSS 等资源：`/en/latest/assets/...`
- Vue Router base：`/en/latest/`
- Fuse 数据：`/en/latest/valaxy-fuse-list.json`

有两点需要注意：

1. 请保留 `base` 末尾的 `/`。VitePress 示例中的 `.replace(/\/$/, '')` 不建议原样照搬；Valaxy 当前的 Fuse 路径会直接拼接 `base`，去掉末尾斜杠后会变成 `/en/latestvalaxy-fuse-list.json`。
2. `public` 目录资源以及 Markdown 中手写的根绝对路径（如 `/image.png`、`/guide/`）不会全部自动补上 `base`，请改用相对路径或自行拼接 base。

`READTHEDOCS_CANONICAL_URL` 确实是 Read the Docs 自动注入的变量，格式包含语言和版本路径，例如 `https://docs.example.com/en/latest/`：https://docs.readthedocs.com/platform/latest/reference/environment-variables.html#envvar-READTHEDOCS_CANONICAL_URL

另外，Read the Docs 的自定义构建还需要把 Valaxy 的 `dist` 内容复制到 `$READTHEDOCS_OUTPUT/html/`；这与 `base` 配置是两件事。

**Reply by doudou0720** [reply]
> public 目录资源以及 Markdown 中手写的根绝对路径（如 /image.png、/guide/）不会全部自动补上 base，请改用相对路径或自行拼接 base。

如果这一个能和vitepress对齐就好了...

**Reply by YunYouJun** [reply]
@doudou0720 已在 [v1.0.0-rc.4](https://github.com/YunYouJun/valaxy/releases/tag/v1.0.0-rc.4) 与 VitePress 对齐，默认生效，不需要额外开关：

- Vite 最终解析的 `base` 会传给页面、摘要/路由和本地搜索的 Markdown 渲染链路；
- Markdown 中的 `/guide/`、`/file.pdf` 等根绝对链接会补上 `base`；
- Markdown 图片交给 Vue/Vite 静态资源链路处理，最终产物会应用 `base`；
- 客户端新增 `withBase()`，供主题配置和 Vue 动态 URL 使用。

外部 URL 和相对路径保持不变；原生 HTML `<a>` 仍可作为显式绕过方式。可通过 `pnpm up valaxy@latest` 升级。部署说明：https://valaxy.site/zh/guide/deploy#deploy-under-base-path
